api/v1alpha1/rollouthistory_types.go

```diff
@@ -0,0 +1,206 @@
+	Service *corev1.Service `json:"service,omitempty"`
+}
+
+// TrafficRoutingWrapper indicates information of Gateway API or Ingress
+type TrafficRoutingWrapper struct {
+	// Ingress indicates information of ingress
+	// +optional
+	IngressWrapper *IngressWrapper `json:"ingressWrapper,omitempty"`
+	// HTTPRouteWrapper indacates information of Gateway API
+	// +optional
+	HTTPRouteWrapper *HTTPRouteWrapper `json:"httpRouteWrapper,omitempty"`
+}
+
+// IngressWrapper indicates information of the ingress related
+type IngressWrapper struct {
+	// Name indicates the ingress name
+	Name string `json:"name,omitempty"`
+	// Ingress indicates the ingress
+	Ingress *networkingv1.Ingress `json:"ingress,omitempty"`
+}
+
+// HTTPRouteWrapper indicates information of gateway API
+type HTTPRouteWrapper struct {
+	//Name indicates the httproute name
+	Name string `json:"name,omitempty"`
+	// HTTPRoute indicates the HTTPRoute
+	HTTPRoute *v1alpha2.HTTPRoute `json:"httpRoute,omitempty"`
+}
+
+// Workload indicates information of the workload, such as cloneset, deployment, advanced statefulset
+type Workload struct {
+	metav1.TypeMeta `json:",inline"`
+	// Name indicates the workload name
+	Name string `json:"name,omitempty"`
+	// Label selector for pods.
+	// It must match the pod template's labels.
+	Selector *metav1.LabelSelector `json:"selector" protobuf:"bytes,2,opt,name=selector"`
+	// Number of desired pods. This is a pointer to distinguish between explicit
+	// zero and not specified. Defaults to 1.
+	// +optional
+	Replicas *int32 `json:"replicas,omitempty" protobuf:"varint,1,opt,name=replicas"`
+	// Type of deployment or CloneSetUpdateStrategy. Can be "Recreate" or "RollingUpdate".
+	// +optional
+	Type string `json:"type,omitempty" protobuf:"bytes,1,opt,name=type,casttype=DeploymentStrategyType"`
+	// Template describes the pods that will be created.
+	// +kubebuilder:pruning:PreserveUnknownFields
+	// +kubebuilder:validation:Schemaless
+	Template v1.PodTemplateSpec `json:"template"`
+}
+
+// RolloutHistoryStatus defines the observed state of RolloutHistory
+type RolloutHistoryStatus struct {
+	// INSERT ADDITIONAL STATUS FIELD - define observed state of cluster
+	// Important: Run "make" to regenerate code after modifying this file
+
+	// Phase indicates phase of RolloutHistory, such as "pending", "progressing", "completed"
+	Phase string `json:"phase,omitempty"`
+	// CanaryStepIndex indicates the current step
+	CanaryStepIndex *int32 `json:"canaryStepIndex,omitempty"`
+	// CanaryStepState indicates state of this rollout revision, such as "init", "pending", "update", "terminated", "completed", "cancelled", whick is upon rollout canary_step_state
+	CanaryStepState string `json:"canaryStepState,omitempty"`
+	// RolloutState indicates the rollouts status
+	RolloutState RolloutState `json:"rolloutState,omitempty"`
+	// canaryStepPods indicates the pods released
+	CanaryStepPods []CanaryStepPods `json:"canaryStepPods,omitempty"`
```

Comment:
CanarySteps []CanaryStep `json:"canarySteps,omitempty"`

type CanaryStep struct {
   // CanaryStepIndex indicates the current step
   CanaryStepIndex int32 `json:"canaryStepIndex,omitempty"`
   // Pods indicates the pods information
   Pods []Pod `json:"pods,omitempty"`
}